bson/bsoncodec/codec_cache.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (C) MongoDB, Inc. 2017-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package bsoncodec
+
+import (
+	"reflect"
+	"sync"
+	"sync/atomic"
+)
+
+// statically assert array size
+var _ = (kindEncoderCache{}).entries[reflect.UnsafePointer]
+var _ = (kindDecoderCache{}).entries[reflect.UnsafePointer]
+
+type typeEncoderCache struct {
+	cache sync.Map // map[reflect.Type]ValueEncoder
+}
+
+func (c *typeEncoderCache) Store(rt reflect.Type, enc ValueEncoder) {
+	c.cache.Store(rt, enc)
+}
+
+func (c *typeEncoderCache) Load(rt reflect.Type) (ValueEncoder, bool) {
+	if v, _ := c.cache.Load(rt); v != nil {
+		return v.(ValueEncoder), true
+	}
+	return nil, false
+}
+
+func (c *typeEncoderCache) LoadOrStore(rt reflect.Type, enc ValueEncoder) ValueEncoder {
+	if v, loaded := c.cache.LoadOrStore(rt, enc); loaded {
+		enc = v.(ValueEncoder)
+	}
+	return enc
+}
+
+func (c *typeEncoderCache) Clone() *typeEncoderCache {
+	cc := new(typeEncoderCache)
+	c.cache.Range(func(k, v interface{}) bool {
+		if k != nil && v != nil {
+			cc.cache.Store(k, v)
+		}
+		return true
+	})
+	return cc
+}
+
+type typeDecoderCache struct {
+	cache sync.Map // map[reflect.Type]ValueDecoder
+}
+
+func (c *typeDecoderCache) Store(rt reflect.Type, dec ValueDecoder) {
+	c.cache.Store(rt, dec)
+}
+
+func (c *typeDecoderCache) Load(rt reflect.Type) (ValueDecoder, bool) {
+	if v, _ := c.cache.Load(rt); v != nil {
+		return v.(ValueDecoder), true
+	}
+	return nil, false
+}
+
+func (c *typeDecoderCache) LoadOrStore(rt reflect.Type, dec ValueDecoder) ValueDecoder {
+	if v, loaded := c.cache.LoadOrStore(rt, dec); loaded {
+		dec = v.(ValueDecoder)
+	}
+	return dec
+}
+
+func (c *typeDecoderCache) Clone() *typeDecoderCache {
+	cc := new(typeDecoderCache)
+	c.cache.Range(func(k, v interface{}) bool {
+		if k != nil && v != nil {
+			cc.cache.Store(k, v)
+		}
+		return true
+	})
+	return cc
+}
+
+// atomic.Value requires that all calls to Store() have the same concrete type
+// so we wrap the ValueEncoder with a kindEncoderCacheEntry to ensure the type
+// is always the same (since different concrete types may implement the
+// ValueEncoder interface).
+type kindEncoderCacheEntry struct {
+	enc ValueEncoder
+}
+
+type kindEncoderCache struct {
+	entries [reflect.UnsafePointer + 1]atomic.Value // *kindEncoderCacheEntry
```

Comment:
I can go either way with this since we quickly store the result returned by the `kind{Encoder,Decoder}Cache` into `Registry.type{Encoders,Decoders}` (honestly, I wish we actually didn't need this and had a better way of handling `lookupInterfaceEncoder`). 

That said, it is a lot faster when benchmarked in isolation and we do test against a new type being added, which shouldn't be possible until Go 2.0 and the last time `reflect.Kind` was [modified](https://go-review.googlesource.com/c/go/+/341333) it did not change the Kind value.

```
$ go test -run '^$' -bench 'BenchmarkKindEncoderCacheLoad' -v
goos: darwin
goarch: arm64
pkg: go.mongodb.org/mongo-driver/bson/bsoncodec
BenchmarkKindEncoderCacheLoad
BenchmarkKindEncoderCacheLoad/Array
BenchmarkKindEncoderCacheLoad/Array-10         	541177398	         2.033 ns/op
BenchmarkKindEncoderCacheLoad/SyncMap
BenchmarkKindEncoderCacheLoad/SyncMap-10       	82671672	        14.64 ns/op
BenchmarkKindEncoderCacheLoad/Map
BenchmarkKindEncoderCacheLoad/Map-10           	207793422	         5.779 ns/op
PASS
ok  	go.mongodb.org/mongo-driver/bson/bsoncodec	4.638s
```

```go
func BenchmarkKindEncoderCacheLoad(b *testing.B) {
	m := new(sync.Map)
	x := make(map[reflect.Kind]ValueEncoder)
	c := new(kindEncoderCache)
	for k := reflect.Kind(0); k <= reflect.UnsafePointer; k++ {
		c.Store(k, fakeCodec{})
		m.Store(k, fakeCodec{})
		x[k] = fakeCodec{}
	}
	b.Run("Array", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			kind := reflect.Kind(i) % (reflect.UnsafePointer + 1)
			_, ok := c.Load(kind)
			if !ok {
				b.Fatal("missing:", kind)
			}
		}
	})
	b.Run("SyncMap", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			kind := reflect.Kind(i) % (reflect.UnsafePointer + 1)
			_, ok := m.Load(kind)
			if !ok {
				b.Fatal("missing:", kind)
			}
		}
	})
	b.Run("Map", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			kind := reflect.Kind(i) % (reflect.UnsafePointer + 1)
			_, ok := x[kind]
			if !ok {
				b.Fatal("wat")
			}
		}
	})
}
```

Reply:
Yeah, `[...]sync.Atomic` definitely benchmarks faster in isolation, and it does seem unlikely that Go 1.x will add new `Kind`s anytime soon. I also see the test now. I'm fine keeping it as-is.

Reply:
My concern is that a user uses an old driver against a new Go with a reflect kind greater than `UnsafePointer`. A user can hardly be cautious of this issue. I would suggest using a `sync.Map` as long as it performs better than the current implementation. Otherwise, we can move the capacity check into an `init()` in "codec_cache.go", like:
```
func init() {
	rt := reflect.UnsafePointer
	s := rt.String()
	if !strings.Contains(s, strconv.Itoa(int(rt))) {
		panic("the capacity of kindEncoderCache is too small")
	}
}
```

Reply:
Fixed with: https://github.com/mongodb/mongo-go-driver/pull/1313/commits/924008c295ec27619785aee96f38f91646eaa009